content/en/docs/tasks/configure-pod-container/resize-container-resources.md

```diff
@@ -0,0 +1,259 @@
+---
+title: Resize CPU and Memory Resources assigned to Containers
+content_type: task
+weight: 30
+---
+
+
+<!-- overview -->
+
+This page assumes that you are familiar with [Quality of Service](/docs/tasks/configure-pod-container/quality-service-pod/)
+for Kubernetes Pods.
+
+This page shows how to resize CPU and memory resources assigned to containers
+of a running pod without restarting the pod or its containers. A Kubernetes node
+allocates resources for a pod based on its **requests**, and limits the pod's
+resource usage based on the **limits** specified in the pod's containers.
+
+Container's resource **requests** and **limits** are now **mutable** for CPU
+and memory resources. They now represent **desired** CPU and memory values.
```

Comment:
> They now represents desired CPU and memory values.

I don't agree to this statement. This `resources` specification has always been the **desired** state. The whole manifest for a Pod is about its **desired** state. IMO, the change introduced in 1.27 is that these `resources` (especially, CPU and memory) are **mutable without having to recreate the Pod or the container**. They used to be mutable, but you have to recreate the Pod.

Reply:
I see your point (and I'm looking for better way to [get the message across](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1287-in-place-update-pod-resources#api-changes)).

The notion of Spec being **desired** state is, in theory, correct. However, in practice, since this field has been immutable for so long (recreating = terminate + create, aka 'not mutable'), there is concern that people (tools) are relying on it as an indicator of **actual** state.

Perhaps a better wording would be `They are now purely a declaration, and represent the most recently desired values for CPU and memory.` ?

Reply:
Emm...

I don't think the fact that both `requests` and `limits` are used to describe the **desired** state has changed. They were and they still are about **desired** state.

They were mutable and they are still mutable today. That is not a change.

The change that really matters (to me) is that they are **mutable in-place**, combined.

Reply:
They are not mutable (i.e. they cannot be updated ... like the image field for e.g. which can be updated on the currently homed pod, and that triggers container restart). 
If you try to update resources with this featuregate disabled, apiserver throws a validation error. The only way you could change them is to create a new pod with new resources and evict the old one. Scheduler is involved in finding a new home for that new pod with new resources. So .. replacement yes, mutation no.

```bash
root@vbuild:~/go/src/k8s.io/kubernetes# k get pod bpod -ojson | jq .spec.containers[0].resources
{
  "limits": {
    "cpu": "500m",
    "memory": "500Mi"
  },
  "requests": {
    "cpu": "300m",
    "memory": "200Mi"
  }
}
root@vbuild:~/go/src/k8s.io/kubernetes# ./cluster/kubectl.sh patch pod bpod --patch '{"spec":{"containers":[{"name":"ctr", "resources":{"requests":{"cpu":"
350m"}}}]}}'
The Pod "bpod" is invalid: spec: Forbidden: pod updates may not change fields other than `spec.containers[*].image`,`spec.initContainers[*].image`,`spec.activeDeadlineSeconds`,`spec.tolerations` (only additions to existing tolerations),`spec.terminationGracePeriodSeconds` (allow it to be set to 1 if it was previously negative)
  core.PodSpec{
  	Volumes:        {{Name: "kube-api-access-txss6", VolumeSource: {Projected: &{Sources: {{ServiceAccountToken: &{ExpirationSeconds: 3607, Path: "token"}}, {ConfigMap: &{LocalObjectReference: {Name: "kube-root-ca.crt"}, Items: {{Key: "ca.crt", Path: "ca.crt"}}}}, {DownwardAPI: &{Items: {{Path: "namespace", FieldRef: &{APIVersion: "v1", FieldPath: "metadata.namespace"}}}}}}, DefaultMode: &420}}}},
  	InitContainers: nil,
  	Containers: []core.Container{
  		{
  			... // 6 identical fields
  			EnvFrom: nil,
  			Env:     nil,
  			Resources: core.ResourceRequirements{
  				Limits: {s"cpu": {i: {...}, s: "500m", Format: "DecimalSI"}, s"memory": {i: {...}, s: "500Mi", Format: "BinarySI"}},
  				Requests: core.ResourceList{
- 					s"cpu":    {i: resource.int64Amount{value: 300, scale: -3}, s: "300m", Format: "DecimalSI"},
+ 					s"cpu":    {i: resource.int64Amount{value: 350, scale: -3}, s: "350m", Format: "DecimalSI"},
  					s"memory": {i: {...}, Format: "BinarySI"},
  				},
  				Claims: nil,
  			},
  			ResizePolicy: nil,
  			VolumeMounts: {{Name: "kube-api-access-txss6", ReadOnly: true, MountPath: "/var/run/secrets/kubernetes.io/serviceaccount"}},
  			... // 12 identical fields
  		},
  	},
  	EphemeralContainers: nil,
  	RestartPolicy:       "OnFailure",
  	... // 28 identical fields
  }

root@vbuild:~/go/src/k8s.io/kubernetes#
```

Reply:
Thanks for the confirmation, especially the `kubectl patch` test!!!

So ... the real change is that the `resources` is now mutable, and the mutation doesn't necessarily cause a Pod restart. Right?

Reply:
Yes. Restarting a container upon mutation of resources can be controlled by policy. Setting the policy to `NotRequired` does not guarantee restart-free resize. K8s may still restart if that's the only way to recover from an odd future failure (not something we do today) or the runtime does not support etc.

Reply:
So does the above suggested https://github.com/kubernetes/website/pull/39845#discussion_r1142816218 wording sound better?

Reply:
I still believe 'They now represent **desired** CPU and memory values.' can be removed safely without causing confusion.

Reply:
SG. Took it out and reformatted to use bullet points for better clarity. thanks.